README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# CTCF

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
<!-- [![BioC status](http://www.bioconductor.org/shields/build/release/bioc/CTCF.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/CTCF) [![R-CMD-check-bioc](https://github.com/mdozmorov/CTCF/actions/workflows/R-CMD-check-bioc.yaml/badge.svg)](https://github.com/mdozmorov/CTCF/actions/workflows/R-CMD-check-bioc.yaml) -->

*[CTCF](https://bioconductor.org/packages/3.16/CTCF)* defines an
AnnotationHub resource representing genomic coordinates of
[FIMO](https://meme-suite.org/meme/doc/fimo.html)-predicted CTCF binding
sites for human and mouse genomes, including the
[Telomere-to-Telomere](https://github.com/marbl/CHM13) and
[mm39](http://genomeref.blogspot.com/2020/07/grcm39-new-mouse-reference-genome.html)
genome assemblies. It also includes experimentally defined CTCF-bound
cis-regulatory elements from [ENCODE
SCREEN](https://screen.encodeproject.org/).

TL;DR - for human hg38 genome assembly, use `hg38.MA0139.1.RData`
(“AH104729”). For mouse mm10 genome assembly, use `mm10.MA0139.1.RData`
(“AH104755”). For [ENCODE SCREEN](https://screen.encodeproject.org/)
data, use `hg38.SCREEN.GRCh38_CTCF.RData` (“AH104730”) or
`mm10.SCREEN.mm10_CTCF.RData` (“AH104756”) objects.

The CTCF GRanges are named as `<assembly>.<Database>`. The
FIMO-predicted data includes extra columns with motif name, score,
p-value, q-value, and the motif sequence.

<!--
**Please, note that the updated CTCF objects will be available in 
Bioconductor/AnnotationHub 3.16.** To test the following code, use the 
`bioconductor::devel` Docker image. Run:

```bash
docker run -e PASSWORD=password -p 8787:8787 -d --rm -v $(pwd):/home/rstudio bioconductor/bioconductor_docker:devel
```
Open http://localhost:8787 and login using `rstudio/password` credentials.
-->

## Installation instructions

[Install](https://www.bioconductor.org/install/#install-R) the latest
release of R, then get the latest version of Bioconductor by starting R
and entering the commands:

``` r
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install(version = "3.16")
```

Then, install additional packages using the following code:

``` r
# BiocManager::install("AnnotationHub", update = FALSE) 
# BiocManager::install("GenomicRanges", update = FALSE)
# BiocManager::install("plyranges", update = FALSE)
```

## Example

``` r
suppressMessages(library(AnnotationHub))
ah <- AnnotationHub()
#> snapshotDate(): 2022-10-31
query_data <- subset(ah, preparerclass == "CTCF")
# Explore the AnnotationHub object
query_data
#> AnnotationHub with 51 records
#> # snapshotDate(): 2022-10-31
#> # $dataprovider: JASPAR 2022, CTCFBSDB 2.0, SwissRegulon, Jolma 2013, HOCOMO...
#> # $species: Homo sapiens, Mus musculus
#> # $rdataclass: GRanges
#> # additional mcols(): taxonomyid, genome, description,
#> #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#> #   rdatapath, sourceurl, sourcetype 
#> # retrieve records with, e.g., 'object[["AH104716"]]' 
#> 
#>              title                                                 
#>   AH104716 | T2T.CIS_BP_2.00_Homo_sapiens.RData                    
#>   AH104717 | T2T.CTCFBSDB_PWM.RData                                
#>   AH104718 | T2T.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData     
#>   AH104719 | T2T.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData
#>   AH104720 | T2T.Jolma2013.RData                                   
#>   ...        ...                                                   
#>   AH104762 | mm9.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData
#>   AH104763 | mm9.Jolma2013.RData                                   
#>   AH104764 | mm9.MA0139.1.RData                                    
#>   AH104765 | mm9.SwissRegulon_human_and_mouse.RData                
#>   AH104766 | mm8.CTCFBSDB.CTCF_predicted_mouse.RData
# Get the list of data providers
query_data$dataprovider %>% table()
#> .
#>           CIS-BP     CTCFBSDB 2.0 ENCODE SCREEN v3     HOCOMOCO v11 
#>                6               12                2                6 
#>      JASPAR 2022       Jolma 2013     SwissRegulon 
#>               13                6                6
```

We can find CTCF sites identified using JASPAR 2022 database in hg38
human genome

``` r
subset(query_data, species == "Homo sapiens" & 
                   genome == "hg38" & 
                   dataprovider == "JASPAR 2022")
#> AnnotationHub with 2 records
#> # snapshotDate(): 2022-10-31
#> # $dataprovider: JASPAR 2022
#> # $species: Homo sapiens
#> # $rdataclass: GRanges
#> # additional mcols(): taxonomyid, genome, description,
#> #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#> #   rdatapath, sourceurl, sourcetype 
#> # retrieve records with, e.g., 'object[["AH104727"]]' 
#> 
#>              title                                                  
#>   AH104727 | hg38.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData
#>   AH104729 | hg38.MA0139.1.RData
# Same for mm10 mouse genome
# subset(query_data, species == "Mus musculus" & genome == "mm10" & dataprovider == "JASPAR 2022")
```

The `hg38.JASPAR2022_CORE_vertebrates_non_redundant_v2` object contains
CTCF sites detected using the [all three CTCF
PWMs](https://jaspar.genereg.net/search?q=CTCF&collection=all&tax_group=all&tax_id=9606&type=all&class=all&family=all&version=all).
To retrieve, we’ll use:

``` r
# hg38.JASPAR2022_CORE_vertebrates_non_redundant_v2
CTCF_hg38_all <- query_data[["AH104727"]]
#> loading from cache
#> require("GenomicRanges")
CTCF_hg38_all
#> GRanges object with 3093041 ranges and 5 metadata columns:
#>             seqnames            ranges strand |                   name
#>                <Rle>         <IRanges>  <Rle> |            <character>
#>         [1]     chr1       11212-11246      + | JASPAR2022_CORE_vert..
#>         [2]     chr1       11399-11432      + | JASPAR2022_CORE_vert..
#>         [3]     chr1       11414-11432      + | JASPAR2022_CORE_vert..
#>         [4]     chr1       12373-12406      + | JASPAR2022_CORE_vert..
#>         [5]     chr1       13507-13541      + | JASPAR2022_CORE_vert..
#>         ...      ...               ...    ... .                    ...
#>   [3093037]     chrY 57215115-57215148      - | JASPAR2022_CORE_vert..
#>   [3093038]     chrY 57215146-57215164      - | JASPAR2022_CORE_vert..
#>   [3093039]     chrY 57215146-57215179      - | JASPAR2022_CORE_vert..
#>   [3093040]     chrY 57215332-57215366      - | JASPAR2022_CORE_vert..
#>   [3093041]     chrY 57216319-57216352      - | JASPAR2022_CORE_vert..
#>                 score    pvalue    qvalue               sequence
#>             <numeric> <numeric> <numeric>            <character>
#>         [1]   7.77064  5.25e-05     0.459 gtgctgtgccagggcgcccc..
#>         [2]  18.48780  2.54e-07     0.118 cagcacgcccacctgctggc..
#>         [3]   9.11475  5.65e-05     0.555    ctggcagctggggacactg
#>         [4]   9.21951  5.25e-05     0.421 CAGCAGGTCTGGCTTTGGCC..
#>         [5]   9.71560  2.24e-05     0.397 GTGCCCTTCCTTTGCTCTGC..
#>         ...       ...       ...       ...                    ...
#>   [3093037]   8.06504  9.11e-05     0.614 CTGCTGGGCCCTCTTGCTCC..
#>   [3093038]   9.11475  5.65e-05     0.726    CTGGCAGCTGGGGACACTG
#>   [3093039]  17.91870  3.72e-07     0.246 CAGCACGCCCGCCTGCTGGC..
#>   [3093040]   7.77064  5.25e-05     0.584 GTGCTGTGCCAGGGCGCCCC..
#>   [3093041]   8.63415  6.96e-05     0.595 CTGCATTTGCGTTCCGACGC..
#>   -------
#>   seqinfo: 24 sequences from hg38 genome
```

The `hg38.MA0139.1` object contains CTCF sites detected using the most
popular [MA0139.1](https://jaspar.genereg.net/matrix/MA0139.1/) CTCF
PWM. To retrieve:

``` r
# hg38.MA0139.1
CTCF_hg38 <- query_data[["AH104729"]]
#> loading from cache
CTCF_hg38
#> GRanges object with 887980 ranges and 5 metadata columns:
#>            seqnames            ranges strand |        name     score    pvalue
#>               <Rle>         <IRanges>  <Rle> | <character> <numeric> <numeric>
#>        [1]     chr1       11414-11432      + |    MA0139.1   9.11475  5.65e-05
#>        [2]     chr1       14316-14334      + |    MA0139.1   7.83607  9.71e-05
#>        [3]     chr1       15439-15457      + |    MA0139.1   8.00000  9.08e-05
#>        [4]     chr1       16603-16621      + |    MA0139.1   8.04918  8.89e-05
#>        [5]     chr1       16651-16669      + |    MA0139.1  11.42620  1.97e-05
#>        ...      ...               ...    ... .         ...       ...       ...
#>   [887976]     chrY 57209918-57209936      - |    MA0139.1  11.42620  1.97e-05
#>   [887977]     chrY 57209966-57209984      - |    MA0139.1   8.04918  8.89e-05
#>   [887978]     chrY 57211133-57211151      - |    MA0139.1   8.00000  9.08e-05
#>   [887979]     chrY 57212256-57212274      - |    MA0139.1   7.83607  9.71e-05
#>   [887980]     chrY 57215146-57215164      - |    MA0139.1   9.11475  5.65e-05
#>               qvalue            sequence
#>            <numeric>         <character>
#>        [1]     0.555 ctggcagctggggacactg
#>        [2]     0.601 GGACCAACAGGGGCAGGAG
#>        [3]     0.599 TAGCCTCCAGAGGCCTCAG
#>        [4]     0.597 CCACCTGAAGGAGACGCGC
#>        [5]     0.504 TGGCCTACAGGGGCCGCGG
#>        ...       ...                 ...
#>   [887976]     0.648 TGGCCTACAGGGGCCGCGG
#>   [887977]     0.770 CCACCTGAAGGAGACGCGC
#>   [887978]     0.770 TAGCCTCCAGAGGCCTCAG
#>   [887979]     0.770 GGACCAACAGGGGCAGGAG
#>   [887980]     0.726 CTGGCAGCTGGGGACACTG
#>   -------
#>   seqinfo: 24 sequences from hg38 genome
```

It is always advisable to sort GRanges objects and keep standard
chromsomes:

``` r
suppressMessages(library(plyranges))
CTCF_hg38_all <- CTCF_hg38_all %>% keepStandardChromosomes() %>% sort()
CTCF_hg38 <- CTCF_hg38 %>% keepStandardChromosomes() %>% sort()
```

Save the data in a BED file, if needed.

``` r
# Note that rtracklayer::import and rtracklayer::export perform unexplained
# start coordinate conversion, likely related to 0- and 1-based coordinate
# system. We recommend converting GRanges to a data frame and save tab-separated
write.table(CTCF_hg38_all %>% sort() %>% as.data.frame(), 
            file = "CTCF_hg38_all.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(CTCF_hg38 %>% sort() %>% as.data.frame(), 
            file = "CTCF_hg38.bed",
            sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

Create an [IGV](https://software.broadinstitute.org/software/igv/) XML
session file out of the saved BED files using the
*[tracktables](https://bioconductor.org/packages/3.16/tracktables)*
package. See `vignette("tracktables", package = "tracktables")` for more
details.

``` r
library(tracktables) # BiocManager::install("tracktables")
# Sample sheet metadata
SampleSheet <- data.frame(SampleName = c("CTCF all", "CTCF MA0139.1"),
                          Description = c("All CTCF matrices from JASPAR2022",
                                          "MA0139.1 CTCF matrix from JASPAR2022"))
# File sheet linking files with sample names
FileSheet <- data.frame(SampleName = c("CTCF all", "CTCF MA0139.1"),
                        bigwig = c(NA, NA),
                        interval = c("CTCF_hg38_all.bed", "CTCF_hg38.bed"),
                        bam = c(NA, NA))
# Creating an IGV session XML file
MakeIGVSession(SampleSheet, FileSheet, 
               igvdirectory = getwd(), "CTCF_from_JASPAR2022", "hg38")
```

Note that the [FIMO](https://meme-suite.org/meme/doc/fimo.html) tool
detects CTCF binding sites using the 1e-4 p-value threshold by default
(the more significant p-value corresponds to the more confidently
detected CTCF motif). We found that this threshold may be too
permissive. Using the [ENCODE SCREEN](https://screen.encodeproject.org/)
database as ground truth, we found 1e-6 as the optimal threshold
providing approximately 80% true positive rate. However, less
significant CTCF motifs may be cell type-specific or have weaker CTCF
binding and therefore be missed by conventional peak callers. If cell
type-specific CTCF binding is of interest, we recommend exploring less
significant CTCF sites.

<img src="man/figures/Figure_human_pvalues_threshold.png" width="100%" />

To filter the GRanges object and keep high-confidence CTCF sites, use:

``` r
# Check length before filtering
print(paste("Number of CTCF motifs at the default 1e-4 threshold:", length(CTCF_hg38)))
#> [1] "Number of CTCF motifs at the default 1e-4 threshold: 887980"
# Filter and check length after filtering
CTCF_hg38_filtered <- CTCF_hg38 %>% plyranges::filter(pvalue < 1e-6)
print(paste("Number of CTCF motifs at the 1e-6 threshold:", length(CTCF_hg38_filtered)))
#> [1] "Number of CTCF motifs at the 1e-6 threshold: 21671"
# Similarly, filter
CTCF_hg38_all_filtered <- CTCF_hg38_all %>% plyranges::filter(pvalue < 1e-6)
```

Given some databases provide multiple CTCF PWMs, one CTCF site may be
detected multiple times resulting in overlapping CTCF sites. For
example, the proportion of overlapping CTCF sites in the
`CTCF_hg38_all_filtered` object containing CTCF sites detected by three
matrices nearly 40%:

``` r
# Proportion of overlapping enrtries
tmp <- findOverlaps(CTCF_hg38_all, CTCF_hg38_all)
prop_overlap <- sort(table(queryHits(tmp)) %>% table(), decreasing = TRUE)
sum(prop_overlap[which(names(prop_overlap) != "1")]) / length(CTCF_hg38_all)
#> [1] 0.375093
```

The proportion of overlapping CTCF sites in the `CTCF_hg38_filtered`
object containing CTCF sites detected by the
[MA0139.1](https://jaspar.genereg.net/matrix/MA0139.1/) matrix is less
than 2.5%

``` r
tmp <- findOverlaps(CTCF_hg38, CTCF_hg38)
prop_overlap <- sort(table(queryHits(tmp)) %>% table(), decreasing = TRUE)
sum(prop_overlap[which(names(prop_overlap) != "1")]) / length(CTCF_hg38)
#> [1] 0.0233485
```

Reducing them (merging overlapping CTCF sites), combined with 1E-6
cutoff filtering, yields the number of CTCF sites comparable to
previously reported.

``` r
print(paste("Number of CTCF_hg38 motifs at the 1e-6 threshold AND reduced:", length(CTCF_hg38_filtered %>% reduce())))
#> [1] "Number of CTCF_hg38 motifs at the 1e-6 threshold AND reduced: 21652"
print(paste("Number of CTCF_hg38_all motifs at the 1e-6 threshold AND reduced:", length(CTCF_hg38_all_filtered %>% reduce())))
#> [1] "Number of CTCF_hg38_all motifs at the 1e-6 threshold AND reduced: 63572"
```

However, regulatory elements with CTCF proteins co-occupying
adjacent/overlapping CTCF binding motifs were shown to be functionally
and structurally different from those with single CTCF motifs. We
provide non-reduced CTCF data and advise considering overlap of CTCF
sites depending on the study’s goal.

# liftOver of CTCF coordinates


NEWS.md
# CTCF 0.99.11 (2022-12-24)
- Update README and vignette, creating IGV session XML

# CTCF 0.99.10 (2022-11-06)
- Update README and vignette, installation instructions

# CTCF 0.99.9 (2022-10-29)
- Update README and vignette

# CTCF 0.99.8 (2022-10-28)
- Update README and vignette

# CTCF 0.99.7 (2022-10-15)
- Update README and vignette: installation instructions, examples of filtering
  by p-value, merging, liftover

# CTCF 0.99.5 (2022-07-20)
- Major update
  + New 51 GRanges objects with predicted and experimental CTCF sites
    + PWM data: JASPAR, HOCOMOCO, CIS-BP, CTCFBSDB, SwissRegulon, Jolma 2013
    + Experimental/predicted data: ENCODE SCREEN, CTCFBSDB
    + Human/mouse genome assemblies, including mm39 and T2T

# CTCF 0.99.4 (2021-06-09)
- Minor fixes
  + Move AnnotationHub to Suggests
  + Remove RefManager package

# CTCF 0.99.2 (2021-06-02)
- Minor fixes
  + Remove LazyData: true
  + Remove GitHub installation instructions
  + Remove link to the Google Drive folder
  + Remove timing of vignette building
  + Add AnnotatiohHub as an imported package

# CTCF 0.99.1 (2021-05-19)
- Minor fixes
  + biocViews edits
  + metadata updated
  + dev folder and R function placeholder removed

# CTCF 0.99.0 (2021-05-08)
- Initial commit


inst/scripts/make-metadata.R
### =========================================================================
### CTCF metadata
### -------------------------------------------------------------------------
###

meta <- data.frame(
    Title = c("T2T.CIS_BP_2.00_Homo_sapiens.RData", "T2T.CTCFBSDB_PWM.RData", "T2T.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData", "T2T.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "T2T.Jolma2013.RData", "T2T.MA0139.1.RData", "T2T.SwissRegulon_human_and_mouse.RData", "hg38.CIS_BP_2.00_Homo_sapiens.RData", "hg38.CTCFBSDB.CTCF_predicted_human.RData", "hg38.CTCFBSDB_PWM.RData", "hg38.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData", "hg38.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "hg38.Jolma2013.RData", "hg38.MA0139.1.RData", "hg38.SCREEN.GRCh38_CTCF.RData", "hg38.SwissRegulon_human_and_mouse.RData", "hg19.CIS_BP_2.00_Homo_sapiens.RData", "hg19.CTCFBSDB.CTCF_predicted_human.RData", "hg19.CTCFBSDB_PWM.RData", "hg19.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData", "hg19.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "hg19.Jolma2013.RData", "hg19.MA0139.1.RData", "hg19.SwissRegulon_human_and_mouse.RData", "hg18.CTCFBSDB.CTCF_predicted_human.RData", "hg18.MA0139.1.RData", "mm39.CIS_BP_2.00_Mus_musculus.RData", "mm39.CTCFBSDB_PWM.RData", "mm39.HOCOMOCOv11_core_MOUSE_mono_meme_format.RData", "mm39.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "mm39.Jolma2013.RData", "mm39.MA0139.1.RData", "mm39.SwissRegulon_human_and_mouse.RData", "mm10.CIS_BP_2.00_Mus_musculus.RData", "mm10.CTCFBSDB.CTCF_predicted_mouse.RData", "mm10.CTCFBSDB_PWM.RData", "mm10.HOCOMOCOv11_core_MOUSE_mono_meme_format.RData", "mm10.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "mm10.Jolma2013.RData", "mm10.MA0139.1.RData", "mm10.SCREEN.mm10_CTCF.RData", "mm10.SwissRegulon_human_and_mouse.RData", "mm9.CIS_BP_2.00_Mus_musculus.RData", "mm9.CTCFBSDB.CTCF_predicted_mouse.RData", "mm9.CTCFBSDB_PWM.RData", "mm9.HOCOMOCOv11_core_MOUSE_mono_meme_format.RData", "mm9.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "mm9.Jolma2013.RData", "mm9.MA0139.1.RData", "mm9.SwissRegulon_human_and_mouse.RData", "mm8.CTCFBSDB.CTCF_predicted_mouse.RData"),
    Description = c("T2T CTCF motifs detected using human PWM matrices from http://cisbp.ccbr.utoronto.ca/, by FIMO",
                    "T2T CTCF motifs detected using PWM matrices from https://insulatordb.uthsc.edu/, by FIMO",
                    "T2T CTCF motifs detected using human PWM matrices from https://hocomoco11.autosome.org/, by FIMO",
                    "T2T CTCF motifs detected using human PWM matrices from https://jaspar.genereg.net/, by FIMO",
                    "T2T CTCF motifs detected using PWM matrices from DOI:10.1016/j.cell.2012.12.009, by FIMO",
                    "T2T CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "T2T CTCF motifs detected using PWM matrices from https://swissregulon.unibas.ch/sr/, by FIMO",
                    "hg38 CTCF motifs detected using human PWM matrices from http://cisbp.ccbr.utoronto.ca/, by FIMO",
                    "hg38 CTCF predicted motifs from https://insulatordb.uthsc.edu/",
                    "hg38 CTCF motifs detected using PWM matrices from https://insulatordb.uthsc.edu/, by FIMO",
                    "hg38 CTCF motifs detected using human PWM matrices from https://hocomoco11.autosome.org/, by FIMO",
                    "hg38 CTCF motifs detected using human PWM matrices from https://jaspar.genereg.net/, by FIMO",
                    "hg38 CTCF motifs detected using PWM matrices from DOI:10.1016/j.cell.2012.12.009, by FIMO",
                    "hg38 CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "hg38 CTCF-bound human cis-regulatory elements from https://screen.encodeproject.org/",
                    "hg38 CTCF motifs detected using PWM matrices from https://swissregulon.unibas.ch/sr/, by FIMO",
                    "hg19 CTCF motifs detected using human PWM matrices from http://cisbp.ccbr.utoronto.ca/, by FIMO",
                    "hg19 CTCF motifs detected using PWM matrices from https://insulatordb.uthsc.edu/, by FIMO",
                    "hg19 CTCF predicted motifs from https://insulatordb.uthsc.edu/",
                    "hg19 CTCF motifs detected using human PWM matrices from https://hocomoco11.autosome.org/, by FIMO",
                    "hg19 CTCF motifs detected using human PWM matrices from https://jaspar.genereg.net/, by FIMO",
                    "hg19 CTCF motifs detected using PWM matrices from DOI:10.1016/j.cell.2012.12.009, by FIMO",
                    "hg19 CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "hg19 CTCF motifs detected using PWM matrices from https://swissregulon.unibas.ch/sr/, by FIMO",
                    "hg18 CTCF predicted motifs from https://insulatordb.uthsc.edu/",
                    "hg18 CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "mm39 CTCF motifs detected using mouse PWM matrices from http://cisbp.ccbr.utoronto.ca/, by FIMO",
                    "mm39 CTCF predicted motifs from https://insulatordb.uthsc.edu/",
                    "mm39 CTCF motifs detected using mouse PWM matrices from https://hocomoco11.autosome.org/, by FIMO",
                    "mm39 CTCF motifs detected using mouse PWM matrices from https://jaspar.genereg.net/, by FIMO",
                    "mm39 CTCF motifs detected using PWM matrices from DOI:10.1016/j.cell.2012.12.009, by FIMO",
                    "mm39 CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "mm39 CTCF motifs detected using PWM matrices from https://swissregulon.unibas.ch/sr/, by FIMO",
                    "mm10 CTCF motifs detected using mouse PWM matrices from http://cisbp.ccbr.utoronto.ca/, by FIMO",
                    "mm10 CTCF motifs detected using PWM matrices from https://insulatordb.uthsc.edu/, by FIMO",
                    "mm10 CTCF predicted motifs from https://insulatordb.uthsc.edu/",
                    "mm10 CTCF motifs detected using mouse PWM matrices from https://hocomoco11.autosome.org/, by FIMO",
                    "mm10 CTCF motifs detected using mouse PWM matrices from https://jaspar.genereg.net/, by FIMO",
                    "mm10 CTCF motifs detected using PWM matrices from DOI:10.1016/j.cell.2012.12.009, by FIMO",
                    "mm10 CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "mm10 CTCF-bound mouse cis-regulatory elements from https://screen.encodeproject.org/",
                    "mm10 CTCF motifs detected using PWM matrices from https://swissregulon.unibas.ch/sr/, by FIMO",
                    "mm9 CTCF motifs detected using mouse PWM matrices from http://cisbp.ccbr.utoronto.ca/, by FIMO",
                    "mm9 CTCF motifs detected using PWM matrices from https://insulatordb.uthsc.edu/, by FIMO",
                    "mm9 CTCF predicted motifs from https://insulatordb.uthsc.edu/",
                    "mm9 CTCF motifs detected using mouse PWM matrices from https://hocomoco11.autosome.org/, by FIMO",
                    "mm9 CTCF motifs detected using mouse PWM matrices from https://jaspar.genereg.net/, by FIMO",
                    "mm9 CTCF motifs detected using PWM matrices from DOI:10.1016/j.cell.2012.12.009, by FIMO",
                    "mm9 CTCF motifs detected using MA0139.1 PWM matrix from https://jaspar.genereg.net/, by FIMO",
                    "mm9 CTCF motifs detected using PWM matrices from https://swissregulon.unibas.ch/sr/, by FIMO",
                    "mm8 CTCF predicted motifs from https://insulatordb.uthsc.edu/"),
    BiocVersion = rep("3.15", 51),
    Genome = c("T2T", "T2T", "T2T", "T2T", "T2T", "T2T", "T2T", "hg38", "hg38", "hg38", "hg38", "hg38", "hg38", "hg38", "hg38", "hg38", "hg19", "hg19", "hg19", "hg19", "hg19", "hg19", "hg19", "hg19", "hg18", "hg18", "mm39", "mm39", "mm39", "mm39", "mm39", "mm39", "mm39", "mm10", "mm10", "mm10", "mm10", "mm10", "mm10", "mm10", "mm10", "mm10", "mm9", "mm9", "mm9", "mm9", "mm9", "mm9", "mm9", "mm9", "mm8"),
    SourceType = rep("RData", 51),
    SourceUrl = rep("https://drive.google.com/drive/folders/19ZXr7IETfks0OdYlmuc1Hqe700Pw3jPc?usp=sharing", 51),
    SourceVersion = rep("July 20 2022", 51),
    Species = c("Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Homo sapiens", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus", "Mus musculus"),
    TaxonomyId = c(9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 9606, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090, 10090),
    Coordinate_1_based = rep(TRUE, 51),
    DataProvider = c("CIS-BP", "CTCFBSDB 2.0", "HOCOMOCO v11", "JASPAR 2022", "Jolma 2013", "JASPAR 2022", "SwissRegulon", "CIS-BP", "CTCFBSDB 2.0", "CTCFBSDB 2.0", "HOCOMOCO v11", "JASPAR 2022", "Jolma 2013", "JASPAR 2022", "ENCODE SCREEN v3", "SwissRegulon", "CIS-BP", "CTCFBSDB 2.0", "CTCFBSDB 2.0", "HOCOMOCO v11", "JASPAR 2022", "Jolma 2013", "JASPAR 2022", "SwissRegulon", "CTCFBSDB 2.0", "JASPAR 2022", "CIS-BP", "CTCFBSDB 2.0", "HOCOMOCO v11", "JASPAR 2022", "Jolma 2013", "JASPAR 2022", "SwissRegulon", "CIS-BP", "CTCFBSDB 2.0", "CTCFBSDB 2.0", "HOCOMOCO v11", "JASPAR 2022", "Jolma 2013", "JASPAR 2022", "ENCODE SCREEN v3", "SwissRegulon", "CIS-BP", "CTCFBSDB 2.0", "CTCFBSDB 2.0", "HOCOMOCO v11", "JASPAR 2022", "Jolma 2013", "JASPAR 2022", "SwissRegulon", "CTCFBSDB 2.0"),
    Maintainer = rep("Mikhail Dozmorov <mikhail.dozmorov@gmail.com>", 51),
    RDataClass = rep("GRanges", 51),
    DispatchClass = rep("Rda", 51),
    RDataPath = c("CTCF/T2T.CIS_BP_2.00_Homo_sapiens.RData", "CTCF/T2T.CTCFBSDB_PWM.RData", "CTCF/T2T.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData", "CTCF/T2T.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "CTCF/T2T.Jolma2013.RData", "CTCF/T2T.MA0139.1.RData", "CTCF/T2T.SwissRegulon_human_and_mouse.RData", "CTCF/hg38.CIS_BP_2.00_Homo_sapiens.RData", "CTCF/hg38.CTCFBSDB.CTCF_predicted_human.RData", "CTCF/hg38.CTCFBSDB_PWM.RData", "CTCF/hg38.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData", "CTCF/hg38.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "CTCF/hg38.Jolma2013.RData", "CTCF/hg38.MA0139.1.RData", "CTCF/hg38.SCREEN.GRCh38_CTCF.RData", "CTCF/hg38.SwissRegulon_human_and_mouse.RData", "CTCF/hg19.CIS_BP_2.00_Homo_sapiens.RData", "CTCF/hg19.CTCFBSDB.CTCF_predicted_human.RData", "CTCF/hg19.CTCFBSDB_PWM.RData", "CTCF/hg19.HOCOMOCOv11_core_HUMAN_mono_meme_format.RData", "CTCF/hg19.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "CTCF/hg19.Jolma2013.RData", "CTCF/hg19.MA0139.1.RData", "CTCF/hg19.SwissRegulon_human_and_mouse.RData", "CTCF/hg18.CTCFBSDB.CTCF_predicted_human.RData", "CTCF/hg18.MA0139.1.RData", "CTCF/mm39.CIS_BP_2.00_Mus_musculus.RData", "CTCF/mm39.CTCFBSDB_PWM.RData", "CTCF/mm39.HOCOMOCOv11_core_MOUSE_mono_meme_format.RData", "CTCF/mm39.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "CTCF/mm39.Jolma2013.RData", "CTCF/mm39.MA0139.1.RData", "CTCF/mm39.SwissRegulon_human_and_mouse.RData", "CTCF/mm10.CIS_BP_2.00_Mus_musculus.RData", "CTCF/mm10.CTCFBSDB.CTCF_predicted_mouse.RData", "CTCF/mm10.CTCFBSDB_PWM.RData", "CTCF/mm10.HOCOMOCOv11_core_MOUSE_mono_meme_format.RData", "CTCF/mm10.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "CTCF/mm10.Jolma2013.RData", "CTCF/mm10.MA0139.1.RData", "CTCF/mm10.SCREEN.mm10_CTCF.RData", "CTCF/mm10.SwissRegulon_human_and_mouse.RData", "CTCF/mm9.CIS_BP_2.00_Mus_musculus.RData", "CTCF/mm9.CTCFBSDB.CTCF_predicted_mouse.RData", "CTCF/mm9.CTCFBSDB_PWM.RData", "CTCF/mm9.HOCOMOCOv11_core_MOUSE_mono_meme_format.RData", "CTCF/mm9.JASPAR2022_CORE_vertebrates_non_redundant_v2.RData", "CTCF/mm9.Jolma2013.RData", "CTCF/mm9.MA0139.1.RData", "CTCF/mm9.SwissRegulon_human_and_mouse.RData", "CTCF/mm8.CTCFBSDB.CTCF_predicted_mouse.RData"),
    Tags = rep("CTCF", 51),
    Notes = "")

write.csv(meta, file="inst/extdata/metadata.csv", row.names=FALSE)
